src/fork_gc.c

```diff
@@ -322,177 +352,193 @@ static void ForkGc_CollectGarbageFromTagIdx(ForkGCCtx *gc, RedisSearchCtx *sctx)
-  blockModified->numBlocksBefore = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  Buffer *b = &blockModified->blk.buf;
-  b->data = ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &b->offset);
+static void FGC_recvRepairedBlock(ForkGC *gc, MSG_RepairedBlock *binfo) {
+  FGC_recvFixed(gc, binfo, sizeof(*binfo));
+  Buffer *b = &binfo->blk.buf;
+  b->data = FGC_recvBuffer(gc, &b->offset);
   b->cap = b->offset;
 }
 
-static bool ForkGc_ReadInvertedIndexFromFork(ForkGCCtx *gc, ForkGc_InvertedIndexData *idxData) {
-  long long blocksRepaired = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  if (!blocksRepaired) {
+static bool FGC_recvInvIdx(ForkGC *gc, InvIdxBuffers *bufs, MSG_IndexInfo *info) {
+  long long wasRepaired = FGC_recvLongLong(gc);
+  if (!wasRepaired) {
     return false;
   }
-
-  idxData->originalSize = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  idxData->blockArrayExists = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  if (idxData->blockArrayExists) {
-    idxData->newBlocksArray =
-        (IndexBlock *)ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &idxData->newBlocksArraySize);
-    idxData->newBlocksArraySize /= sizeof(IndexBlock);
+  FGC_recvFixed(gc, info, sizeof(*info));
+  bufs->newBlocklist = FGC_recvBuffer(gc, &bufs->newBlocklistSize);
+  if (bufs->newBlocklistSize) {
+    bufs->newBlocklistSize /= sizeof(*bufs->newBlocklist);
   }
 
-  idxData->freeBufs = (char **)ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &idxData->freeBufsSize);
-  idxData->freeBufsSize /= sizeof(char *);
-
-  long long blocksModifiedSize = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  idxData->bytesCollected = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  idxData->docsCollected = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  idxData->blocksModified = array_new(ModifiedBlock, blocksModifiedSize);
-  for (int i = 0; i < blocksModifiedSize; ++i) {
-    ModifiedBlock mb;
-    ForkGc_ReadModifiedBlock(gc, &mb);
-    idxData->blocksModified = array_append(idxData->blocksModified, mb);
+  bufs->delBlocks = FGC_recvBuffer(gc, &bufs->numDelBlocks);
+  bufs->numDelBlocks /= sizeof(*bufs->delBlocks);
+  bufs->changedBlocks = rm_malloc(sizeof(*bufs->changedBlocks) * info->nblocksRepaired);
+  for (size_t i = 0; i < info->nblocksRepaired; ++i) {
+    FGC_recvRepairedBlock(gc, bufs->changedBlocks + i);
   }
   return true;
 }
 
-static void ForkGc_FixInvertedIndex(ForkGCCtx *gc, ForkGc_InvertedIndexData *idxData,
-                                    InvertedIndex *idx) {
-  if (idxData->freeBufs) {
-    for (int i = 0; i < idxData->freeBufsSize; ++i) {
-      rm_free(idxData->freeBufs[i]);
-    }
-    rm_free(idxData->freeBufs);
+static void checkLastBlock(ForkGC *gc, InvIdxBuffers *idxData, MSG_IndexInfo *info,
+                           InvertedIndex *idx) {
+  IndexBlock *lastOld = idx->blocks + info->nblocksOrig - 1;
+  if (info->lastblkDocsRemoved == 0) {
+    // didn't touch last block in child
+    return;
```

Comment:
If the last block was not touched by the child you are still adding it to the block list and if the parent did touch it then it will be override. Am I missing something?

Reply:
I do not think you cover this in the tests, pls do.

Reply:
Fixed it in the tests, please have a look.